experiments/idris/src/Fathom/Base.idr

```diff
@@ -0,0 +1,107 @@
+module Fathom.Base
+
+
+import Data.Colist
+import Data.List
+
+
+------------------
+-- USEFUL TYPES --
+------------------
+
+
+||| A value that is refined by a proposition.
+|||
+||| This is a bit like `(x : A ** B)`, but with the second element erased.
+public export
+record Refine (0 A : Type) (0 P : A -> Type) where
+  constructor MkRefine
+  ||| The wrapped value
+  value : A
+  ||| The proof of the proposition
+  0 prf : P value
+
+||| Refine a value with a proposition
+public export
+refine : {0 A : Type} -> {0 P : A -> Type} -> (value : A) -> {auto 0 prf : P value} -> Refine A P
+refine value {prf} = MkRefine { value, prf }
+
+
+||| Singleton types
+|||
+||| Inspired by [this type](https://agda.readthedocs.io/en/v2.5.4.1/language/with-abstraction.html#the-inspect-idiom)
+||| from the Agda docs.
+public export
+data Sing : {0 A : Type} -> (x : A) -> Type where
+  MkSing : {0 A : Type} -> {0 x : A} -> (y : A) -> {auto prf : x = y} -> Sing x
```

Comment:
As strongly as I dislike Agda’s generalisation of lower case names in type signatures, it’s kind of forcing my hand due to the lack of sugar for dependent function types, or the ability to drop annotations for implicit arguments. This would be more cleanly defined as:

```idris
data Sing : (x : a) -> Type where
  MkSing : (y : a) -> {auto prf : x = y} -> Sing x
```